workers/unity/Packages/io.improbable.gdk.tools/.CodeGenDocs/3-using-the-codewriter.md

```diff
@@ -0,0 +1,826 @@
+
+```csharp
+cw.Namespace("Your.Own.Namespace", ns =>
+{
+    ns.Annotate("Test").Type("public class TypeName", t =>
+    {
+        //define code
+    });
+
+    ns.Annotate("Test").Enum("public enum EnumName", t =>
+    {
+        //define members
+    });
+});
+```
+
+Generated code:
+
+```csharp
+namespace Your.Own.Namespace
+{
+    [Test]
+    public class TypeName
+    {
+        //code
+    }
+
+    [Test]
+    public enum EnumName
+    {
+        //members
+    }
+}
+```
+
+### Inside a type
+
+Within a type, the `Annotate` method will also let you create annotated methods.
+
+```csharp
+cw.Namespace("Your.Own.Namespace", ns =>
+{
+    ns.Type("public class TypeName", t =>
+    {
+        t.Annotate("TestAnnotation").Method("public void TestMethod()", m =>
+        {
+            //define code
+        });
+    });
+});
+```
+
+Generated code:
+
+```csharp
+namespace Your.Own.Namespace
+{
+    public class TypeName
+    {
+        [TestAnnotation]
+        public void TestMethod()
+        {
+            //code
+        }
+    }
+}
+```
+
+## Initializer
+
+An initializer shares similar traits to a custom scope and enums, but the key difference is that it must end with a semi-colon. The `Initializer` API exists to cater for the difference.
+
+### Example: initialize by hand
+
+```csharp
+t.Initializer("private static readonly int[] CountToFive = new int[]", () => new[]
+{
+    "0",
+    "1",
+    "2",
+    "3",
+    "4",
+    "5"
+});
+```
+
+Generated code:
+
+```csharp
+private static readonly int[] CountToFive = new int[]
+{
+    0,
+    1,
+    2,
+    3,
+    4,
+    5
+}
+```
+
+### Example: initialize from list
+
+```csharp
+var zeroToFive = new List<string> { "0", "1", "2", "3", "4", "5" };
+
+...
+
+t.Initializer("private static readonly int[] CountToFive = new int[]", () => zeroToFive);
+```
+
+Generated code:
+
+```csharp
+private static readonly int[] CountToFive = new int[]
+{
+    0,
+    1,
+    2,
+    3,
+    4,
+    5
+}
+```
+
+## Scoped bodies
+
+A scoped body is used to describe the content of scopes that can contain some actual logic. For example, a `Namespace` does not have logic directly within it - you'd have to define logic inside a `Method` within a `Type`.
+
+A `Method` is an example of a scoped body, because you can define arbitrary logic inside it. Examples of scoped bodies include:
+
+* methods (as mentioned)
+* loops
+* each scope within an `if-elseif-else` statement
+* each scope within a `try-catch-finally` statement
+* custom scopes
+
+> In the code generator, an abstract `ScopeBody` class exists to generalise the API available for types of scoped bodies.
+
+The API described in the sections below is avaiable from all scoped bodies.
+
+### Loops
+
+The `Loop` method is a special type of custom scope body that also offers `Continue` and `Break` methods.
+
+#### Example: basic loop
```

Comment:
do we want to put the examples into collapsibles?

Reply:
yes!

Reply:
on second thoughts no, as it makes the page look a bit inconsistent if some sections have collapsibles and others don't